server/api/initial_data.py

```diff
@@ -638,8 +638,8 @@ def _migrate_artifacts_batch(
         last_migrated_artifact_id = artifact.id
 
         # save the artifact's tags and labels to migrate them later
-        tag = artifact_metadata.get("tag", "")
-        if tag:
+        tags = _get_tag_names(db, db_session, artifact, artifact_metadata)
+        for tag in tags:
```

Comment:
You're doing a query on the tag table per artifact - that's extremely non-efficient. Since you're doing this part in batches anyway, I suggest you list once all the tags whose `obj_id` is within the current batch and keep that in-memory, then use this data-structure to get tag names.
The data structure is not going to be exceedingly large since it's just for a batch of objects of limited size, and it will save you thousands of DB queries which are expensive.